### PR TITLE
Resource object instances returned by HydroShare object are memcached

### PR DESCRIPTION
**Pull request by aaraney**
It may be desirable for a `Resource` returned by `HydroShare.resource(...)` be memcached in the `HydroShare` instance. This means any metadata retrieved and cached at the `Resource` object level are maintained in memory through a reference in the `HydroShare` instance. Here, this is implemented through a dictionary, `HydroShare._resource_object_cache`. 

There are two downsides (that I can think of) to this implementation that should be considered:

1. `HydroShare.resource(...)` now is not thread safe. One could argue it was not thread safe previously as it passes a reference to a mutable `HydroShareSession`object, but for the most part, that object should be read only.
2. Introduces complexity in managing references. Meaning, if a user wanted the garbage collector to destroy a `Resource` instantiated via `HydroShare.resource(...)`, they would have to remove the key value pair from the `HydroShare._resource_object_cache` dictionary. I don't think this is really an issue, but I wanted to mention it as it is an assumption the software is now imposing.

### CODE REVIEW

**Comment by aaraney** [comment]
Tests failed because hydroshare credentials were not passed as env variables.

**Comment by sblack-usu** [comment]
I'll take a look at what went wrong with the tests when I get a chance.  Thanks for submitting this so quickly.

**Comment by aaraney** [comment]
@sblack-usu are the secrets used in the `gh-action` repo level secrets? I wonder if i'm added in a different role to the project if it will fix this.

**Comment by aaraney** [comment]
@sblack-usu this appears to be the issue. Under a repos settings/secrets:

Secrets are not passed to workflows that are triggered by a pull request from a fork. [Learn more](https://docs.github.com/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets).

**Comment by sblack-usu** [comment]
You're right, that's unfortunate.  I've added you as a collaborator with write access.

**Comment by sblack-usu** [comment]
It's still failing.  I wonder if the PR needs to be resubmitted instead of just redploying the job to pick up the newly available secrets?

**Comment by aaraney** [comment]
I think it is an issue with using a PR from a fork workflow. I just created a new branch on my fork and added a `gh-action` that just runs the unit tests. The branch is based off of 3243379  (the last commit  in the PR at this time).

**Comment by aaraney** [comment]
Tests are still failing, but it doesn't look like anything i've added here:
 
```shell
=========================== short test summary info ============================
FAILED tests/test_functional.py::test_creator_order - AttributeError: 'str' o...
FAILED tests/test_functional.py::test_resource_metadata_updating - AttributeE...
FAILED tests/test_functional.py::test_resource_delete - AssertionError: asser...
FAILED tests/test_functional.py::test_aggregation_fileset[files0] - Exception...
FAILED tests/test_functional.py::test_aggregations[files4] - Exception: Faile...
FAILED tests/test_functional.py::test_filename_spaces - Exception: Failed GET...
ERROR tests/test_functional.py::test_pandas_series_local - Exception: Failed ...
ERROR tests/test_functional.py::test_folder_download - Exception: Failed GET ...
============== 6 failed, 37 passed, 2 errors in 959.08s (0:15:59) ==============
```

**Comment by sblack-usu** [comment]
let's table this for now and I'll take a look soon.

**Comment by sblack-usu** [comment]
@horsburgh - What do you think about caching Resource objects in the HydroShare instance?  @aaraney's implementation looks good and I'm working on figuring out what is going on with the test secrets when a fork submits a PR (the reason why the tests are failing).  The concern over thread safety when accessing/updating a dictionary isn't a concern for me, the dictionary won't get corrupted, the only possible issue is a race condition when saving multiple instances of the same resource to the cache in multiple threads.  It's definitely an edge case and we could solve it with a lock, but the case is rare and the consequence is small.

We could solve @aaraney second point by including a `clear_resource_cache()` method and documenting the feature.

File: hsclient/hydroshare.py
Comment (sblack-usu) [inline hsclient/hydroshare.py:907]: Github doesn't allow me to create a suggestion further above, but the use_cache parameter would also have to be added above.  Then the use_cache method would be checked before reading from and writing to the cache.

```suggestion
        :param validate: Defaults to True, set to False to not validate the resource exists
        :param use_cache: Defaults to True, set to False to skip the cache, and always retrieve the resource from HydroShare.  This parameter also does not cache the retrieved Resource object.
```

Comment (sblack-usu) [reply hsclient/hydroshare.py:907]: A test would have to be added to check this parameter as well.

**Comment by sblack-usu** [comment]
The same cache functions could be added to aggregations within a resource.  I don't think aggregations should hold up this PR though.  @aaraney, do you have an eta for this?  I'm aiming for a release this week if you have the time.

**Comment by aaraney** [comment]
> The same cache functions could be added to aggregations within a resource. I don't think aggregations should hold up this PR though. @aaraney, do you have an eta for this? I'm aiming for a release this week if you have the time.

@sblack-usu, I can add in the requested `use_cache` param plus the units tests pushed up this afternoon. Is there anything else you would like me to include?

**Comment by sblack-usu** [comment]
That should cover it.

**Comment by aaraney** [comment]
@sblack-usu In writing tests for the aforementioned additions, I ran into an issue that requires an implementation decision. I wanted to shed light on it before I go further. 

In short, `HydroShare.create` calls `HydroShare.resource` (with the default `use_cache` parameter), so if a user creates a resource from an instantiated `HydroShare` object, they will implicitly add the resource to the resource cache. This already existed for the `validate` kwarg as well. What are your thoughts?

**Comment by sblack-usu** [comment]
@aaraney - sorry for the slow response.  I suggest adding the use_cache parameter with a default of True to `create` and pass that parameter to `resource`.

**Comment by aaraney** [comment]
> @aaraney - sorry for the slow response. I suggest adding the use_cache parameter with a default of True to create and pass that parameter to resource.

No worries, sorry I am just now getting around to this myself. As you will see above, I added `use_cache` to `HydroShare.create` as requested. With that, are we ready to merge? If you'd like me to bump the minor package version, im happy to do that as well.

**Review by sblack-usu — APPROVED** [review]

**Comment by sblack-usu** [comment]
Thanks @aaraney.  I'll do the minor bump when I do the release.  I really need to update the release action to use the version given with the github release eventually.

**Pull request merged**
